ksqldb-engine/src/main/java/io/confluent/ksql/schema/ksql/inference/SchemaRegisterInjector.java

```diff
@@ -155,80 +167,121 @@ private void registerSchemas(
       final String statementText,
       final boolean registerIfSchemaExists
   ) {
-    registerSchema(
-        schema.key(),
-        kafkaTopic,
-        keyFormat,
-        keySerdeFeatures,
-        config,
-        statementText,
-        registerIfSchemaExists,
-        getSRSubject(kafkaTopic, true),
-        true
-    );
+    if (kvRawSchema.left == null) {
+      registerSchema(
+          schema.key(),
+          kafkaTopic,
+          keyFormat,
+          keySerdeFeatures,
+          config,
+          statementText,
+          registerIfSchemaExists,
+          getSRSubject(kafkaTopic, true),
+          true
+      );
+    } else {
+      registerRawSchema(
+          kvRawSchema.left,
+          kafkaTopic,
+          keyFormat,
+          config,
+          statementText,
+          getSRSubject(kafkaTopic, true),
+          true);
+    }
 
-    registerSchema(
-        schema.value(),
-        kafkaTopic,
-        valueFormat,
-        valueSerdeFeatures,
-        config,
-        statementText,
-        registerIfSchemaExists,
-        getSRSubject(kafkaTopic, false),
-        false
-    );
+    if (kvRawSchema.right == null) {
+      registerSchema(
+          schema.value(),
+          kafkaTopic,
+          valueFormat,
+          valueSerdeFeatures,
+          config,
+          statementText,
+          registerIfSchemaExists,
+          getSRSubject(kafkaTopic, false),
+          false
+      );
+    } else {
+      registerRawSchema(
+          kvRawSchema.right,
+          kafkaTopic,
+          valueFormat,
+          config,
+          statementText,
+          getSRSubject(kafkaTopic, false),
+          false);
+    }
   }
 
-  private void registerSchema(
-      final List<? extends SimpleColumn> schema,
+  private void registerRawSchema(
+      final SchemaAndId schemaAndId,
       final String topic,
       final FormatInfo formatInfo,
-      final SerdeFeatures serdeFeatures,
       final KsqlConfig config,
       final String statementText,
-      final boolean registerIfSchemaExists,
       final String subject,
-      final boolean isKey
+      final Boolean isKey
   ) {
+    final String schemaIdPropStr = isKey ? "key_schema_id" : "value_schema_id";
     final Format format = FormatFactory.of(formatInfo);
-    if (!format.supportsFeature(SerdeFeature.SCHEMA_INFERENCE)) {
-      return;
+    if (!canRegister(format, config, topic)) {
+      throw new KsqlStatementException(schemaIdPropStr + " is provided but format "
+          + format.name() + " doesn't support registering in Schema Registry",
+          statementText);
     }
```

Comment:
Shouldn't this check be called earlier in the `registerSchemas` method to avoid registering schemas without completing the create execution?

For instance, say the key_schema_id is registered, but then the value_schema_id fails because the format does not support schema inference. The CREATE will fail leaving a schema registered in SR, right?

I think you should check for schema inference support and other validations earlier before attempting to register all the schemas.

Reply:
Make sense. For original code path, it doesn't throw, but for raw schema exist case, it will throw. I can check before registering for raw schema exist case